ranger/gui/widgets/titlebar.py

```diff
@@ -101,28 +101,40 @@ def _get_left_part(self, bar):
             bar.add(self.fm.hostname, 'hostname', clr, fixed=True)
             bar.add(' ', 'hostname', clr, fixed=True)
 
-        pathway = self.fm.thistab.pathway
-        if self.settings.tilde_in_titlebar \
-           and (self.fm.thisdir.path.startswith(self.fm.home_path + "/")
-                or self.fm.thisdir.path == self.fm.home_path):
-            pathway = pathway[self.fm.home_path.count('/') + 1:]
-            bar.add('~/', 'directory', fixed=True)
-
-        for path in pathway:
-            if path.is_link:
-                clr = 'link'
-            else:
-                clr = 'directory'
-
-            bidi_basename = self.bidi_transpose(path.basename)
-            bar.add(bidi_basename, clr, directory=path)
+        if self.fm.thisdir:
+            pathway = self.fm.thistab.pathway
+            if self.settings.tilde_in_titlebar \
```

Comment:
I assume you've tried reducing the duplication here?

Reply:
Actually no, this change is related to the control of switching to "dead" tabs, more specifically, to avoid errors of calling a non-existent `pathway` attribute for a tab with `thisdir` of type `None` instead of the `Directory` class.

Reply:
Yes, I understand but small parts of the code are duplicated here so I was wondering whether you'd attempted to reduce that?

Reply:
> If you could squash all the reverted commits out of the history and rephrase the "WIP" commit message I think this is ready.

Done!
I rebased the branch by squashing all commits related to closing dead tabs with a confirmation when switching between them since this logic was later discarded.
And improved the commit message for the initial version of dead-path tab switch control.

BTW, thanks for your refactoring. :wink:

Reply:
Another way to deduplicate code (left part of screen - "before", right - "after"):
<img width="1911" alt="2023-04-19_003252" src="https://user-images.githubusercontent.com/63558838/232898890-f3e6dce0-48f9-4df7-b115-072641dea19c.png">
Formally, we get rid of duplicating pieces of code, but we have to check the condition of the external if operator three times, which was previously checked only once.
In my opinion, this construction reads worse, so I would prefer not to push this commit.

Reply:
Up to you, feel free to use your best judgement. Do you still consider this a draft? The changes look fine to me for the most part.

Reply:
By draft, I mean that there are still unresolved drawbacks and limitations in the dead-path tab switching control implementation, which I've described here and in the related commit message.  But as we discussed earlier, it seems that we can merge it in the current state and come back to improve the current implementation in the future.